uri/uri.h

```diff
@@ -36,6 +40,87 @@ struct Uri {
     [[nodiscard]] bool operator==(Uri const &) const = default;
 };
 
+enum HostType { DNS_DOMAIN, IP4_ADDR, IP6_ADDR, OPAQUE, EMPTY };
+
+struct Host {
+    HostType type;
+
+    // union {
+    std::string domain;
+    std::uint32_t ip4_addr;
+    std::uint16_t ip6_addr[8];
+    //};
+};
+
+struct Origin {
+    std::string scheme;
+    Host *host;
+    std::optional<uint16_t> port;
+    std::optional<std::string> domain;
+    // Need this placeholder until I figure out what "opaqueness" means for an origin in this context
+    bool opaque;
+};
+
+enum ParserState {
+    SCHEME_START,
+    SCHEME,
+    NO_SCHEME,
+    SPECIAL_RELATIVE_OR_AUTHORITY,
+    PATH_OR_AUTHORITY,
+    RELATIVE,
+    RELATIVE_SLASH,
+    SPECIAL_AUTHORITY_SLASHES,
+    SPECIAL_AUTHORITY_IGNORE_SLASHES,
+    AUTHORITY,
+    HOST,
+    HOSTNAME,
+    PORT,
+    FILE,
+    FILE_SLASH,
+    FILE_HOST,
+    PATH_START,
+    PATH,
+    OPAQUE_PATH,
+    QUERY,
+    FRAGMENT,
+    FAILURE,
+    TERMINATE
+};
+
+inline const std::map<std::string, std::string> special_schemes = {
```

Comment:
It feels like the URL class and friends isn't really ready to be merged? Lots of random work-in-progress bits, but no real implementation at all and no tests. It's not really possible to give any form of reasonable review when not even a basic vertical slice or at least a spec link showing that the data structure is reasonable is there.

Reply:
We can also discuss design on Tox or IRC if you want. I can't tell you if any of what's in here is reasonable just looking at what we have right now. The API you call to start parsing a url looks like what I started with before you wanted to take over. Very spec-compliant. :P